File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/auth/JcrLoginModule.java
package solutions.digamma.damas.jcr.auth;

import javax.security.auth.Subject;
import javax.security.auth.callback.Callback;
import javax.security.auth.callback.CallbackHandler;
import javax.security.auth.callback.NameCallback;
import javax.security.auth.callback.PasswordCallback;
import javax.security.auth.callback.UnsupportedCallbackException;
import javax.security.auth.login.LoginException;
import javax.security.auth.spi.LoginModule;
import java.io.IOException;
import java.security.Principal;
import java.util.List;
import java.util.Map;

/**
 * @author Ahmad Shahwan
 */
public class JcrLoginModule implements LoginModule {

    private Subject subject;
    private CallbackHandler callbackHandler;
    private String login;
    private char[] password;
    private Map<String, ?> sharedState;
    private List<String> roles;

    private static final String USERNAME = "javax.security.auth.login.name";
    private static final String PASSWORD = "javax.security.auth.login.password";

    @Override
    public void initialize(
            Subject subject,
            CallbackHandler callbackHandler,
            Map<String, ?> sharedState,
            Map<String, ?> options) {
        this.subject = subject;
        this.callbackHandler = callbackHandler;
        this.sharedState = sharedState;
        this.extractCredentials();
    }

    @Override
    public boolean login() throws LoginException {
        if (this.login == null || this.password == null) {
            throw new LoginException("Missing login");
        }
        return false;
    }

    @Override
    public boolean commit() throws LoginException {
        return false;
    }

    @Override
    public boolean abort() throws LoginException {
        return false;
    }

    @Override
    public boolean logout() throws LoginException {
        return false;
    }

    private void extractCredentials() {
        Object userObj = sharedState.get(USERNAME);
        Object passObj = sharedState.get(PASSWORD);
        if (userObj instanceof Principal)
            this.login = ((Principal) userObj).getName();
        else if (userObj != null) {
            this.login = userObj.toString();
        }
        if (passObj instanceof char[] )
            this.password = (char[]) passObj;
        else if (passObj != null ) {
            this.password = passObj.toString().toCharArray();
        }
        if (this.login != null && this.password != null) {
            return;
        }
        NameCallback nameCallback = new NameCallback("username");
        PasswordCallback passwordCallback =
                new PasswordCallback("password", false);
        try {
            this.callbackHandler.handle(new Callback[] {
                    nameCallback, passwordCallback
            });
            this.login = nameCallback.getName();
            this.password = passwordCallback.getPassword();
        } catch (IOException | UnsupportedCallbackException e) {
            this.login = null;
            this.password = null;
        }
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/auth/JcrLoginModule.kt
package solutions.digamma.damas.jcr.auth

import java.security.Principal
import javax.security.auth.Subject
import javax.security.auth.callback.CallbackHandler
import javax.security.auth.callback.NameCallback
import javax.security.auth.callback.PasswordCallback
import javax.security.auth.callback.UnsupportedCallbackException
import javax.security.auth.login.LoginException
import javax.security.auth.spi.LoginModule

/**
 * @author Ahmad Shahwan
 */
class JcrLoginModule : LoginModule {

    private var subject: Subject? = null
    private var callbackHandler: CallbackHandler? = null
    private var login: String? = null
    private var password: CharArray? = null
    private var sharedState: Map<String, *>? = null
    private val roles: List<String>? = null

    override fun initialize(
            subject: Subject,
            callbackHandler: CallbackHandler,
            sharedState: Map<String, *>,
            options: Map<String, *>) {
        this.subject = subject
        this.callbackHandler = callbackHandler
        this.sharedState = sharedState
        this.extractCredentials()
    }

    @Throws(LoginException::class)
    override fun login(): Boolean {
        this.login ?: throw LoginException("Missing login")
        this.password ?: throw LoginException("Missing password")
        return false
    }

    @Throws(LoginException::class)
    override fun commit(): Boolean {
        return false
    }

    @Throws(LoginException::class)
    override fun abort(): Boolean {
        return false
    }

    @Throws(LoginException::class)
    override fun logout(): Boolean {
        return false
    }

    private fun extractCredentials() {
        val userObj = sharedState!![USERNAME]
        val passObj = sharedState!![PASSWORD]
        if (userObj is Principal)
            this.login = userObj.name
        else if (userObj != null) {
            this.login = userObj.toString()
        }
        if (passObj is CharArray)
            this.password = passObj
        else if (passObj != null) {
            this.password = passObj.toString().toCharArray()
        }
        if (this.login != null && this.password != null) {
            return
        }
        val nameCallback = NameCallback("username")
        val passwordCallback = PasswordCallback("password", false)
        try {
            this.callbackHandler!!.handle(
                    arrayOf(nameCallback, passwordCallback))
            this.login = nameCallback.name
            this.password = passwordCallback.password
        } catch (e: UnsupportedCallbackException) {
            this.login = null
            this.password = null
        }
    }

    companion object {

        private val USERNAME = "javax.security.auth.login.name"
        private val PASSWORD = "javax.security.auth.login.password"
    }
}
